x/ibc/light-clients/07-tendermint/types/proposal_handle.go

```diff
@@ -12,116 +10,114 @@ import (
+	ctx sdk.Context, cdc codec.BinaryMarshaler, subjectClientStore,
+	substituteClientStore sdk.KVStore, substituteClient exported.ClientState,
+	initialHeight exported.Height,
+) (exported.ClientState, error) {
+	substituteClientState, ok := substituteClient.(*ClientState)
 	if !ok {
-		return nil, nil, sdkerrors.Wrapf(
-			clienttypes.ErrInvalidHeader, "expected type %T, got %T", &Header{}, header,
+		return nil, sdkerrors.Wrapf(
+			clienttypes.ErrInvalidClient, "expected type %T, got %T", &ClientState{}, substituteClient,
 		)
 	}
 
+	if !IsMatchingClientState(cs, *substituteClientState) {
+		return nil, sdkerrors.Wrap(clienttypes.ErrInvalidSubstitute, "subject client state does not match substitute client state")
+	}
+
 	// get consensus state corresponding to client state to check if the client is expired
-	consensusState, err := GetConsensusState(clientStore, cdc, cs.GetLatestHeight())
+	consensusState, err := GetConsensusState(subjectClientStore, cdc, cs.GetLatestHeight())
 	if err != nil {
-		return nil, nil, sdkerrors.Wrapf(
-			err, "could not get consensus state from clientstore at height: %d", cs.GetLatestHeight(),
+		return nil, sdkerrors.Wrapf(
+			err, "unexpected error: could not get consensus state from clientstore at height: %d", cs.GetLatestHeight(),
 		)
 	}
 
 	switch {
 
 	case cs.IsFrozen():
 		if !cs.AllowUpdateAfterMisbehaviour {
-			return nil, nil, sdkerrors.Wrap(clienttypes.ErrUpdateClientFailed, "client is not allowed to be unfrozen")
+			return nil, sdkerrors.Wrap(clienttypes.ErrUpdateClientFailed, "client is not allowed to be unfrozen")
 		}
 
 		// unfreeze the client
 		cs.FrozenHeight = clienttypes.ZeroHeight()
 
-		// if the client is expired we unexpire the client using softer validation, otherwise
-		// full validation on the header is performed.
-		if cs.IsExpired(consensusState.Timestamp, ctx.BlockTime()) {
-			return cs.unexpireClient(ctx, clientStore, consensusState, tmHeader, ctx.BlockTime())
+	case cs.IsExpired(consensusState.Timestamp, ctx.BlockTime()):
+		if !cs.AllowUpdateAfterExpiry {
+			return nil, sdkerrors.Wrap(clienttypes.ErrUpdateClientFailed, "client is not allowed to be unexpired")
 		}
 
-		// NOTE: the client may be frozen again since the misbehaviour evidence may
-		// not be expired yet
-		return cs.CheckHeaderAndUpdateState(ctx, cdc, clientStore, header)
-
-	case cs.AllowUpdateAfterExpiry && cs.IsExpired(consensusState.Timestamp, ctx.BlockTime()):
-		return cs.unexpireClient(ctx, clientStore, consensusState, tmHeader, ctx.BlockTime())
-
 	default:
-		return nil, nil, sdkerrors.Wrap(clienttypes.ErrUpdateClientFailed, "client cannot be updated with proposal")
+		return nil, sdkerrors.Wrap(clienttypes.ErrUpdateClientFailed, "client cannot be updated with proposal")
 	}
 
-}
+	// copy consensus states and processed time from substitute to subject
+	// starting from initial height and ending on the latest height (inclusive)
+	// CONTRACT: the revision number is same for substitute and subject
```

Comment:
Why do they need to be in the same revision?

If the substitute client is at a different revision, then just fill in the consensus states at that revision number. We don't need to fill in the revision heights of earlier revisions, we can leave them blank. Because nothing in the IBC protocol (except for planned upgrades which is irrelevant here) relies on making a proof at a particular height,

Reply:
The concern was originally when an upgrade occurs for the substitute during the voting period, see [ref](https://github.com/cosmos/cosmos-sdk/pull/8405#issuecomment-764783975). If this occurs, I don't know if we can safely determine at which height the upgrade occurred since I'm not certain the ordering from a range query would be correct. 

Given that now the chain-id can be different, I think I can remove the check in 02-client, but add a check in 07-tendermint that the initial height revision number and the substitute client revision number are equal. Disallowing using substitute clients that have been upgraded. If an upgrade happens to occur, a new governance proposal could be made using an initial height after the upgrade